### ISSUE TITLE
MPI Fortran Problem (OpenMPI-2.1.1 from OpenFOAM)

### ISSUE THREAD

**Issue by glanzkaiser** [open-mpi/ompi#10477]
Hi all,

I am trying to compile Trilinos with OpenMPI-2.1.1 from OpenFOAM and I get this:

![Capture d’écran_2022-06-17_15-11-23](https://user-images.githubusercontent.com/72222484/174258859-75da8655-bc65-4d0e-af87-ab555c03a237.png)

Then I try to compile SuperLUDist and I get this:

![Capture d’écran_2022-06-17_15-24-44](https://user-images.githubusercontent.com/72222484/174258985-7abad7b2-34b9-40e5-9ded-89e61c4b4076.png)

**My previous trial and error:**
I am trying to build Trilinos  with MPI. I installed OpenFOAM-9 with its 3rd party Library including OpenMPI 2.1.1.

Then I try Trilinos example and it stated it needs libmpi.so.40 which means it needs OpenMPI-4.x.x

I install OpenMPI-4.1.4, I delete the OpenMPI-2.1.1 but when I type: **mpicc --showme** it still look for configuration file from OpenMPI-2.1.1 / **mpicc --showme still pointing to OpenMPI 2** even if I have two OpenMPI (version 4 and 2).

The most problematic is when trying "Hello World" for MPI:
1. Works when using OpenMPI-2.1.1
2. Not working with any OpenMPI version 4

**OpenMPI-4 is superseeded by OpenMPI-2**

If anyone knows how to make sure the command for mpicc mpiCC mpirun to look only at the installed path of OpenMPI-4.1.4 ?

Thank You

**I have put explanation here (opening an issue at Trilinos related to MPI error): [Trilinos-issue](https://github.com/trilinos/Trilinos/issues/10582)**

Picture attached below: [mpicc error](https://i.stack.imgur.com/hpZ5M.png)

**Comment by ggouaillardet** [comment]
Please copy/paste the error message instead of posting screenshots.

is there a link between Trillions and OpenFOAM?

is there any reason why you want Trillionos to use Open FOAM embedded (and old) Open MPI 2.1.1?
you claim you deleted Open MPI 2.1.1, but `mpicc --showme` still refers 2.1.1, which is very odd (are you sure you deleted correctly and how?) `libmpi_mpifh.so` is the Open MPI library for the Fortran bindings, it is possible the embedded MPI was configure'd with `--disable-mpi-fortran` and hence did not build that library.

**Comment by glanzkaiser** [comment]
Sorry for that, I have a hard time to copy/paste the error message since it is in Qemu. 

The link is both are science packages that I want to use, and the reason is because OpenMPI 2.1.1 works for "Hello World" and I want to use that for MPI enabling in building Trilinos. Instead of using new OpenMPI.

I understand that it is most likely OpenFOAM setting for OpenMPI 2.1.1 is disabling-mpi-fortran.

How I deleted OpenMPI 2.1.1
I delete the whole files related by using:
**rm -rf**

then I adjust the path to the OpenMPI-4 files (bin/lib/include) but still when I tried "Hello World" example I cannot use OpenMPI-4.

These are the errors when trying to run Hello World example for MPI or examples for Trilinos by using OpenMPI4++:
* An error occured in MPI_init
* On a NULL communicator
* opal_init failed
* ompi_mpi_init: ompi_rte_init failed
* opal_shmem_base_select failed -> Returned value -1 instead of OPAL_SUCCESS

This is the screenshot:
![blfs-prob-38-1](https://user-images.githubusercontent.com/72222484/174306810-65d1991e-4abf-4bb5-8d57-325246cd2400.png)

**Comment by jsquyres** [comment]
It looks like there is still an Open MPI v2.x install somewhere on your system; I suspect that the problems you have running Open MPI v4.x. is because there is an unintentional mixing of Open MPI v2.x and v4.x on your system, and things go downhill from there.

You didn't use our template for filing Open MPI issues; can you please answer these questions:

* Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
* If you are building/installing from a git clone, please copy-n-paste the output from `git submodule status`.
* Please describe the system on which you are running
    * Operating system/version: 
    * Computer hardware: 
    * Network type:

**Comment by glanzkaiser** [comment]
How to uninstall the OpenMPI-2 completely? 

Here you go:
1. Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
 
For OpenMPI-4.1.3 and OpenMPI-4.1.4 I download it here:
[ompidownload](https://www.open-mpi.org/software/ompi/v4.1/)
using the tarball then extract it then build it and install it

For OpenMPI-2
it is ThirdParty Library from OpenFOAM-9

2. If you are building/installing from a git clone, please copy-n-paste the output from git submodule status.

Never try using git clone. Is it recommended to git clone OMPI? 
    
3. Please describe the system on which you are running
        Operating system/version: a GNU/Linux OS I created from Linux From Scratch 11.0 book then to BLFS book  -> GFreya OS
        Computer hardware: I use Qemu for this OS, my laptop is Dell Precision 6510 with RAM 16GB and Core i7
        Network type: I use simple indihome wireless for internet, no LAN.

**Comment by jsquyres** [comment]
> 1. Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
> 
> **For OpenMPI-4.1.3 and OpenMPI-4.1.4 I download it here:** [ompidownload](https://www.open-mpi.org/software/ompi/v4.1/) using the tarball then extract it then build it and install it

What was the exact `configure` command line that you used to build Open MPI v4.x?  Did it succeed?  After that, did you build and install Open MPI v4.x successfully?

> **For OpenMPI-2** it is ThirdParty Library from OpenFOAM-9

I cannot speak to how OpenFOAM is bundling Open MPI 2.x; we have no control of how downstream / 3rd-party packages integrate with Open MPI.  You'll have to ask them how to uninstall any version of Open MPI that their build system installed on your system, and how to avoid using it in the future (i.e., so that you can use an externally-built-and-installed Open MPI v4.x).
 
> 2. If you are building/installing from a git clone, please copy-n-paste the output from git submodule status.
> 
> **Never try using git clone. Is it recommended to git clone OMPI?**

No, the official tarball releases (e.g., v4.1.4) are fine.

> 3. Please describe the system on which you are running
>    Operating system/version: **a GNU/Linux OS I created from Linux From Scratch 11.0 book then to BLFS book  -> GFreya OS**
>    Computer hardware: **I use Qemu for this OS, my laptop is Dell Precision 6510 with RAM 16GB and Core i7**
>    =Network type: **I use simple indihome wireless for internet, no LAN.**

When you say you are using Qemu, does this mean you are running Windows as a base OS, and then running your Linux From Scratch install in a Qemu VM?

Regardless, I assume that since you are using Linux From Scratch, that means that there is no OS-provided Open MPI package, and this is why you are downloading the Open MPI v4.1.x tarball and trying to use that.  Is that correct?

**Comment by glanzkaiser** [comment]
What was the exact configure command line that you used to build Open MPI v4.x? Did it succeed? After that, did you build and install Open MPI v4.x successfully?

First I download then enter the directory and compile it:
wget https://download.open-mpi.org/release/open-mpi/v4.1/openmpi-4.1.3.tar.gz
**tar -xvf openmpi-4.1.3.tar.gz
cd openmpi-4.1.3
./configure --prefix=/opt/hamzstlib**
(in /opt/hamzstlib there are folders for bin include lib)
**make -j8**
**make install**

The build is successful 100% and the installation too. But I cannot use Hello World example with OpenMPI-4 
[HellloWorldMPI](https://github.com/glanzkaiser/glanzshamzs/tree/main/MPI)

After configuring
![ompi4-1-3](https://user-images.githubusercontent.com/72222484/174426932-146d1063-a727-4717-99fb-29c37444c0b2.png)

After compiling
![openmpimake](https://user-images.githubusercontent.com/72222484/174428438-e7772f9d-ad48-4841-96e3-8ff0464bf75b.png)

After installing OpenMPI-4.1.3 -> this is the funny problematic when I type mpicc --showme (it does not link to OpenMPI-4.1.3)
![Capture d’écran_2022-06-18_14-56-32](https://user-images.githubusercontent.com/72222484/174428623-63363a8b-65b1-4c4d-be18-4da01074ea8e.png)

![Capture d’écran_2022-06-18_14-58-21](https://user-images.githubusercontent.com/72222484/174428642-d3f0c112-d938-4768-9ccc-c6a7e7d6099a.png)

**Comment by ggouaillardet** [comment]
What does `type mpicc` says?

**Comment by glanzkaiser** [comment]
'type mpicc'

mpicc is hashed (/opt/hamzstlib/openfoam/ThirdParty-9/platforms/linux64Gcc/openmpi-2.1.1/bin/mpicc)

**Comment by ggouaillardet** [comment]
it would be funny if it pointed to `/opt/hamzstlib/bin/mpicc`

**Comment by glanzkaiser** [comment]
how to point it to `/opt/hamzstlib/bin/mpicc` ?

**Comment by jsquyres** [comment]
## tl;dr

You should probably start your entire OpenFoam / Open MPI installation over from scratch.  Otherwise, you're likely to accidentally unknowingly re-use incorrectly-installed pieces from your prior attempts.

In doing so, you should probably un-tar/configure/build/install Open MPI v4.x completely from scratch (so that there's no possibility of mixing v2.x and v4.x).  Do this first -- before attempting to re-install OpenFoam.

Then you should consult the OpenFoam docs to understand how to build it with your externally-installed Open MPI v4.x (and not use any OpenFoam-provided copy of Open MPI).

## More detail

A few things I notice from the information you sent:

* I notice that `mpicc` (and friends) are a different color (cyan) than most other executables in `/opt/hamzstlib/bin`.  Does this mean that it is a sym link and not a regular executable file?
    * Further lending credence to that theory, I notice that `/opt/hamzstlib/bin/mpicc --showme` and `mpicc --showme` (which `type mpicc` revealed to be pointing to `/opt/hamzstlib/openfoam/ThirdParty-9/platforms/linux64Gcc/openmpi-2.1.1/bin/mpicc`) showed the same output: both including CLI flags of directories under `/opt/hamzstlib/openfoam/...`.
* It looks like you configured and built Open MPI v4.1.x, but did you install it?  I don't see any mention of you invoking `make install`.
    * Note: this is somewhat moot, per the above points (e.g., if OpenFoam installed its own copy of Open MPI into some deep directory location `/opt/hamzstlib/openfoam/ThirdParty-9/platform/linux64Gcc/openmpi-2.1.1` and then sym linked at least some strategic files back into `/opt/hamzstlib`).
    * Meaning: if you actually `make install` Open MPI v4.1 into `/opt/hamzstlib`, I think you'd have an even bigger mess / muddling of Open MPI v2.x and v4.x than you already [may] have.
* Can you answer my question / verify if my assumption is correct about Windows / Qemu VM? (see previous comment on this github issue)

I still think that the root cause of your issues is that somehow Open MPI v2.x and v4.x have gotten muddled together on your system (perhaps either by mixing the installations themselves, or mixing MPI application executables and Open MPI installations).  It's probably best to not try to fix it -- rather, remove everything and start over so that you can get a good, clean, correct install (rather than getting a broken install and trying to fix it).

**Comment by glanzkaiser** [comment]
**You should probably start your entire OpenFoam / Open MPI installation over from scratch.**
Alright I will uninstall OpenFOAM and ask the developer how to use OpenMPI 4 with OpenFOAM.

**I notice that mpicc (and friends) are a different color (cyan) than most other executables in /opt/hamzstlib/bin. Does this mean that it is a sym link and not a regular executable file?** 

Yes the OpenMPI-4 install the bin in symlinks mode, instead of green color one like the regular/ordinary bin files. 

The mpicc --showme always pointing to the openfoam directory(OpenFOAM has 2 directories: the main openFOAM and the ThirdParty for libraries like OpenMPI, scotch, Paraview, etc).

**It looks like you configured and built Open MPI v4.1.x, but did you install it? I don't see any mention of you invoking make install.** 
I install it, and do `make install`, thus making the symlinks on the /opt/hamzstlib/bin. It is the location for OpenMPI-4 installation.

I hope this OpenMPI will be solved as soon as possible, good luck for me then.

**Comment by jsquyres** [comment]
Ok, good luck.  I'm going to close this issue then (since you're going to go back to the OpenFoam community with the question there).  Open a new issue if a new question arises; thanks.
